app/services/sso_service.rb

```diff
@@ -0,0 +1,112 @@
+# frozen_string_literal: true
+
+require 'saml/auth_fail_handler'
+require 'sentry_logging'
+
+class SSOService
+  include SentryLogging
+  include ActiveModel::Validations
+
+  def initialize(saml_response)
+    raise 'SAML Response is required' if saml_response.nil?
+    @saml_response = saml_response
+    @saml_attributes = SAML::User.new(saml_response)
+    @existing_user = User.find(saml_attributes.user_attributes.uuid)
+    @new_login = @existing_user.present?
+    @new_user_identity = UserIdentity.new(saml_attributes.to_hash)
+    @new_user = init_new_user(new_user_identity, existing_user, saml_attributes.changing_multifactor?)
+    @new_session = Session.new(uuid: new_user.uuid)
+  end
+
+  attr_reader :new_session, :new_user, :new_user_identity, :saml_attributes, :saml_response, :existing_user,
+              :failure_instrumentation_tag
+  # TODO: eventually will rip AuthFailHandler out and make it a custom validator
+  validate :composite_validations
+
+  def self.extend_session!(session, user)
+    session.expire(Session.redis_namespace_ttl)
+    user&.identity&.expire(UserIdentity.redis_namespace_ttl)
+    user&.expire(User.redis_namespace_ttl)
+  end
+
+  def persist_authentication!
+    existing_user.destroy if existing_user.present?
+    if valid?
+      new_session.save && new_user.save && new_user_identity.save
+    else
+      handle_error_reporting_and_instrumentation
+    end
+  end
+
+  def new_login?
+    @new_login
```

Comment:
Maybe this could just be `existing_user.present?` since you don't really make use of `@new_login` otherwise.

Reply:
Or should it be `!existing_user.present?`, since it's used to increment the StatsD for `STATSD_LOGIN_NEW_USER_KEY`?

Reply:
https://github.com/department-of-veterans-affairs/vets-api/pull/1726, i'll make the change you suggested and we can merge this as well.